staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -1071,68 +1095,103 @@ func taintTolerated(taint resourceapi.DeviceTaint, request requestAccessor) bool
-	for _, consumedCounter := range device.basic.ConsumesCounters {
-		referencedSharedCounters.Insert(consumedCounter.CounterSet)
-	}
-
-	// Create a structure that captures the initial counter for all sharedCounters
-	// referenced by the device.
-	availableCounters := make(map[draapi.UniqueString]map[string]draapi.Counter)
-	for _, counterSet := range slice.Spec.SharedCounters {
-		if !referencedSharedCounters.Has(counterSet.Name) {
-			// the API validation logic has been added to make sure the counterSet referred should exist in capacityPools
-			continue
-		}
-		counterShared := make(map[string]draapi.Counter, len(counterSet.Counters))
-		for name, cap := range counterSet.Counters {
-			counterShared[name] = cap
+	sliceName := draapi.MakeUniqueString(slice.Name)
+
+	// Check first if the available counters for this slice have already been
+	// calculated.
+	alloc.mutex.RLock()
+	availableCountersForSlice, found := alloc.availableCounters[sliceName]
+	alloc.mutex.RUnlock()
+	// If not, we need to do it now. But we store the result so it doesn't need
+	// to be calculated again.
+	// Since this is computed without holding the lock on the mutex, other goroutines
+	// might also do this work. But the input will be the same to all of them, so
+	// the result will also always be the same.
+	if !found {
+		availableCountersForSlice = make(counterSets)
+		for _, counterSet := range slice.Spec.SharedCounters {
+			availableCountersForCounterSet := make(map[string]draapi.Counter, len(counterSet.Counters))
+			for name, c := range counterSet.Counters {
+				availableCountersForCounterSet[name] = c
+			}
+			availableCountersForSlice[counterSet.Name] = availableCountersForCounterSet
 		}
-		availableCounters[counterSet.Name] = counterShared
-	}
 
-	// Update the data structure to reflect capacity already in use.
-	for _, device := range slice.Spec.Devices {
-		deviceID := DeviceID{
-			Driver: slice.Spec.Driver,
-			Pool:   slice.Spec.Pool.Name,
-			Device: device.Name,
-		}
-		if !alloc.allocatedDevices.Has(deviceID) && !alloc.allocatingDevices[deviceID] {
-			continue
-		}
-		for _, consumedCounter := range device.Basic.ConsumesCounters {
-			counterShared := availableCounters[consumedCounter.CounterSet]
-			for name, cap := range consumedCounter.Counters {
-				existingCap, ok := counterShared[name]
-				if !ok {
-					// the API validation logic has been added to make sure the capacity referred should exist in capacityPools
-					continue
+		// Update the data structure to reflect counters already consumed by allocated devices.
+		for _, device := range slice.Spec.Devices {
+			deviceID := DeviceID{
+				Driver: slice.Spec.Driver,
+				Pool:   slice.Spec.Pool.Name,
+				Device: device.Name,
+			}
+			// No devices can be in the allocating state when this is computed. The counters
+			// consumed by any devices in the allocating state will be tracked in the
+			// consumedCounters data structure.
```

Comment:
The comment doesn't seem to match the following if check?

I understand what it is saying, but I would put it  last in the for loop as explanation for code that isn't there and first explain what the existing code does.

Reply:
Yeah, I've updated some of the comments to make it clearer.